crates/ra_ide/src/syntax_highlighting/injection.rs

```diff
@@ -137,7 +137,7 @@ pub(super) fn highlight_doc_comment(
     let (analysis, tmp_file_id) = Analysis::from_single_file(text);
 
     stack.push();
-    for mut h in analysis.highlight(tmp_file_id).unwrap() {
+    for mut h in analysis.highlight_injection(tmp_file_id).unwrap() {
```

Comment:
Let's use `super::highlight(&analysis.db, tmp_file_id, None, false)` here to avoid adding a new method to Analysis. 

Also, looking at this I now realize that there's a big problem with cancellation here... Basically, we can't cancel stuff while we are doing highlighting of the injection. Seems bad, long term. 

We need some salsa-level magic to solve this in a nicer way...

Reply:
As an alternative, wouldn't it be better to parse doc comments (maybe including markdown) directly when parsing the file? This would certainly remove a lot of complexity from the current injection logic (and the extraction part) and maybe one can simulate the steps from https://doc.rust-lang.org/rustdoc/documentation-tests.html in order to get a faithful semantic representation.